detect and ignore __might_sleep line number changes

joe-lawrence opened the pull request:
When CONFIG_DEBUG_ATOMIC_SLEEP is enabled, __might_sleep calls will add
the line number to the instruction stream.  Detect and revert any such
changes.

Fixes: #657.

joe-lawrence commented:
Lightly tested on fedora/4.10.  I can throw the test suite at this if it seems like a reasonable approach.  If we get more of these line-number bumps, perhaps we refactor these routines with a generic pattern description parsing?  Didn't seem worth it just yet, so I copy/pasted ```kpatch_warn_only_change()```

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -321,6 +321,93 @@ static int kpatch_warn_only_change(struct section *sec)
+
+		/* verify the next instruction is mov $0x0,%rdi */
+		if (*(unsigned int *)(start1 + offset + length) != 0xc7c748)
+			return 0;
```
Setting CONFIG_DEBUG_ATOMIC_SLEEP and looking through the compiled code at calls to ___might_sleep(), this pattern doesn't always hold.  Sometimes it sets %esi first, sometimes it sets %rdi first, sometimes there are instructions in-between.  So I would remove this restriction.

After doing that, I think the only difference between this function and kpatch_warn_only_change() is the string compares inside the rela loop.  So I think it would be worthwhile to combine the two functions into a single `kpatch_line_macro_change_only()` function.  It might be as simple as adding the "__might_sleep" check at the end of the loop.

joe-lawrence replied:
Thanks for double checking, you're right about the code pattern.
**block/blk-core.c :: generic_make_request_checks()** is an interesting example of this:
```
0000000000000000 <generic_make_request_checks>:
   0:   0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
                        1: R_X86_64_NONE        __fentry__-0x4
   5:   55                      push   %rbp
   6:   31 d2                   xor    %edx,%edx
>  8:   be 60 07 00 00          mov    $0x760,%esi
   d:   48 89 e5                mov    %rsp,%rbp
  10:   41 57                   push   %r15
  12:   41 56                   push   %r14
  14:   41 55                   push   %r13
  16:   41 54                   push   %r12
  18:   49 89 ff                mov    %rdi,%r15
  1b:   53                      push   %rbx
  1c:   48 83 ec 38             sub    $0x38,%rsp
  20:   8b 5f 28                mov    0x28(%rdi),%ebx
> 23:   48 c7 c7 00 00 00 00    mov    $0x0,%rdi
                        26: R_X86_64_32S        .rodata.blk_start_queue.str1.1
  2a:   65 48 8b 04 25 28 00    mov    %gs:0x28,%rax
  31:   00 00
  33:   48 89 45 d0             mov    %rax,-0x30(%rbp)
  37:   31 c0                   xor    %eax,%eax
> 39:   e8 00 00 00 00          callq  3e <generic_make_request_checks+0x3e>
                        3a: R_X86_64_PC32       __might_sleep-0x4
```
where the line number is stored before saving the stack frame, then the filename relocation is saved, more instructions, then finally the callq + __might_sleep rela.

If I understand **kpatch_warn_only_change()** correctly, it's verifying for a given section:
- before and after versions must match up until...
- an immediate mov to %esi.
- a "warn_slowpath_" non-string relocation somewhere later in this section
(BTW, does checking for the optional relocations matter?  I guess it saves a couple extra strcmp's)

Adding "__might_sleep" to the relocation check works out for **generic_make_request_checks()** but I'm wondering if this algorithm is a bit optimistic.  I'm trying to think of an example patch that could inadvertently pass the test, but it's getting late on Friday afternoon and I'm stumped :)

I'd be curious to see other patterns that you may had noticed.

jpoimboe commented:
> before and after versions must match up until...
> an immediate mov to %esi.

To be clear, it also ensures that the mov to esi is the *only* change in the entire function (unless there are other pattern matches).

> (BTW, does checking for the optional relocations matter? I guess it saves a couple extra strcmp's)

Yes, because it ensures that there are no other relocations in-between the mov to %esi and the call to might_sleep/warn.  Otherwise, for example, the relocation could be for a call to another function, for which %esi would be a function argument.

> Adding "__might_sleep" to the relocation check works out for generic_make_request_checks() but I'm wondering if this algorithm is a bit optimistic.

It's a valid concern.  But IMO it would be highly unlikely for the algorithm to come up with a false positive.  And anyway, as a safeguard, we expect the user to always check the output of kpatch-build to ensure the right functions were detected as changed.

I suppose we could make it a little safer by separating the relocation loop into two separate loops -- one for warn and one for might_sleep.  Not sure if that's really necessary.

> I'd be curious to see other patterns that you may had noticed.

Looking around with `objdump -dr vmlinux.o` and searching for `might_sleep`, I didn't see any other patterns.  Sometimes the esi mov comes first, sometimes the rdi file name string mov comes first.  That's the same reason `kpatch_warn_only_change()` treats the string relocation as optional, since it could have come before the detected esi mov.


> Edit: just noticed that ___might_sleep (three underscores) might need to be checked as well.

Ah, right, both the two- and three-underscore versions should be checked for.

joe-lawrence commented:
Okay, one more time through **kpatch_warn_only_change()** and I understand now.  I wasn't considering the continuation of checks through the rest of the section after a potential WARN macro match. 

I'm running the long kpatch-test to verify no inadvertent hiccups in the test cases.  I'll post the results once that's finished, but in the mean time I'll push v2 as you may have comments on the inline comments or naming conventions.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -302,7 +310,9 @@ static int kpatch_warn_only_change(struct section *sec)
-			if (!strncmp(rela->sym->name, "warn_slowpath_", 14)) {
+			if (!strncmp(rela->sym->name, "warn_slowpath_", 14) ||
+			   (!strncmp(rela->sym->name, "__might_sleep", 14)) ||
+			   (!strncmp(rela->sym->name, "___might_sleep", 15))) {
```
I think just a strcmp compare will suffice since they're not substrings.  Otherwise the patch looks good to me.

joe-lawrence commented:
v3
changes: use **strcmp** instead of **strncmp** as suggested by @jpoimboe 
testing on Fedora 25 + 4.9 +  ```CONFIG_DEBUG_ATOMIC_SLEEP```:

- The long test finished and aside from meminfo-cmdline-rebuild-SLOW.patch, all were happy.  (I was getting a bunch of ```find_local_syms: 113: find_local_syms for xyz.c: found_none``` errors from this patch, but I don't think its related this this PR.)

- From the integration tests, gcc-static-local-var-5.patch contains a few instances of the compact might_sleep pattern.  I verified that the line numbers in **kauditd_thread()** moved, but that create-diff-object ignored the change to that function:

```
Disassembly of section .text.kauditd_thread:

0000000000000000 <kauditd_thread>:
...
 121:	be 23 02 00 00       	mov    $0x223,%esi
 126:	48 c7 c7 00 00 00 00 	mov    $0x0,%rdi
			129: R_X86_64_32S	.rodata.kauditd_thread.str1.1
 12d:	e8 00 00 00 00       	callq  132 <kauditd_thread+0x132>
			12e: R_X86_64_PC32	__might_sleep-0x4

...
checking file kernel/audit.c
Reading special section data
Building original kernel
Building patched kernel
Extracting new and modified ELF sections
audit.o: changed function: audit_log_config_change
audit.o: new function: kpatch_audit_foo
audit.o: new function: kpatch_audit_check
audit.o: changed function: audit_log_end
audit.o: changed function: audit_log_start
Patched objects: vmlinux
Building patch module: kpatch-gcc-static-local-var-5.ko
SUCCESS
```

- Using a custom patch below, I tested the longer version with similar results:

```
diff -Nupr src.orig/block/blk-core.c src/block/blk-core.c
--- src.orig/block/blk-core.c   2017-02-10 14:25:47.844626220 -0500
+++ src/block/blk-core.c        2017-02-10 16:51:11.475236308 -0500
@@ -1885,6 +1885,10 @@ generic_make_request_checks(struct bio *
        char b[BDEVNAME_SIZE];
        struct hd_struct *part;
 
+
+
+
+
        might_sleep();
 
        if (bio_check_eod(bio, nr_sectors))


0000000000000000 <generic_make_request_checks>:
...
   8:   be 60 07 00 00          mov    $0x760,%esi
...
  23:   48 c7 c7 00 00 00 00    mov    $0x0,%rdi
                        26: R_X86_64_32S        .rodata.blk_start_queue.str1.1
...
  39:   e8 00 00 00 00          callq  3e <generic_make_request_checks+0x3e>
                        3a: R_X86_64_PC32       __might_sleep-0x4

...
checking file block/blk-core.c
Reading special section data
Building original kernel
Building patched kernel
Extracting new and modified ELF sections
ERROR: no functional changes found. Check /root/.kpatch/build.log for more details.
```

jpoimboe commented:
Forget to push it? :-)

joe-lawrence commented:
Yup :)  should be up now.

jpoimboe commented:
Thanks @joe-lawrence !

👍

joe-lawrence commented:
>The long test finished and aside from meminfo-cmdline-rebuild-SLOW.patch, all were happy. (I was getting a bunch of find_local_syms: 113: find_local_syms for xyz.c: found_none errors from this patch, but I don't think its related this this PR.)

Hmm, I'm seeing this on a stock RHEL7 box with the same patch from this branch...  It's blowing up on a ton of .o files.  I don't remember ever seeing these error messages before.   Before we merge this PR, this afternoon I'll see if I can repro with kpatch master and figure out what's going on :(

jpoimboe commented:
Ok, thanks for testing it.  You never know what's going to break where.  We really need to get that CI testing going.

joe-lawrence commented:
I'll open a new issue, but the find_local_syms errors I was getting are **not** related to this patchset, so feel to merge.

Another observation: regarding the relatively new ```convert-global-local.patch```, is it expected to see this warning on load:
```
kpatch: Skipped dynrela for this_cpu_cmpxchg16b_emu (0xffffffffc058d527 <- 0xffffffff8f40fec0): the instruction has been changed already.
kpatch: This is not necessarily a bug but it may indicate in some cases that the binary patch does not handle paravirt operations, alternatives or the like properly.
```

jpoimboe commented:
That warnings looks like a known limitation of kpatch.ko, related to the handling of .altinstructions sections.  I think there's nothing we can do about it.  It should be fixed in livepatch though.

👍

flaming-toast commented:
@joe-lawrence Yeah, that's a known issue with applying dynrelas to sections that have been paravirt patched (see arch/x86/kernel/alternative.c) It's something that the kpatch core module cannot handle, because the module loader typically applies these paravirt/alternative sections to the patch module before the kpatch core module gets a chance to apply dynrelas to it. We skip applying the dynrela in this case, because we don't want to overwrite the paravirt patched instruction that had been applied by the module loader.  With livepatch, we got around this issue by renaming these sections with a ".klp.arch." prefix and applying them manually with `arch_klp_init_module_loaded()`.
